### ISSUE TITLE
Support showing deprecated completion items

### ISSUE THREAD

**Issue by yyoncho** [emacs-lsp/lsp-mode#2904]
As it title - we should display in company-mode when the item is deprecated. The first step is to have support for that in company-mode. (cc @dgutov)

**Comment by ericdallo** [comment]
`clojure-lsp` already return the tags for those completion items, we can test with that server when available.

**Comment by dgutov** [comment]
> The first step is to have support for that in company-mode

What kind of support? Do you want them displayed with strikethrough or something?

The cheapest way is to just mention the deprecation in annotation, of course.

**Comment by yyoncho** [comment]
> Do you want them displayed with strikethrough or something? 

strikethrough sounds good.

**Comment by kiennq** [comment]
Or if `company-mode` can preserve the face attribute of the passed items it would also be great. `lsp-mode` can add the strike-through itself. At least, the vanilla `completion-at-point` is preserving the face and can show the strike-through already

**Comment by dgutov** [comment]
> Or if company-mode can preserve the face attribute of the passed items it would also be great.

That would be a breaking change, and not necessarily for the best.

There is a backend action called `pre-render` which allows adding properties to completion strings. It's not piped through `company-capf`, but that could be done.

But the proper thing to do here is probably to add the `company-tooltip-deprecated` face, a `deprecated` backend command and a `company-capf` adapter for it.

**Comment by robertoaloi** [comment]
Hi @yyoncho Is my understanding correct that, before the "strikethrough" is available to users for deprecated functions changes are required to `company-mode` itself? If so, is there a ticket that I can track?

**Comment by yyoncho** [comment]
@robertoaloi it is implemented and merged in company-mode. (As a side note, I have implemented support for deprecated only via the deprecated property `deprecated`, not via the tags).

**Comment by robertoaloi** [comment]
> I have implemented support for deprecated only via the deprecated property deprecated, not via the tags

Ah, that explains it. I am playing around with deprecated items in the Erlang LS, but using tags. They displayed correctly in VS Code, but not in Emacs. Thanks for the quick respose!

**Comment by yyoncho** [comment]
Adding support here should be easy - https://github.com/yyoncho/lsp-mode/blob/warn-home-folder/lsp-completion.el#L322 . If you can provide me with sample server response, I can add it.

**Comment by robertoaloi** [comment]
For a `completionItem`, there will simply be an additional, optional, list of `tags`. The only possible tag for now is `1`, for _deprecated_.

This is according to the [specs](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#completionItem).

So, something like:

```
[Trace - 05:52:57 PM] Received response 'completionItem/resolve - (118)' in 78ms.
Result: {
  "tags": [
    1
  ],
  "label": "parse_query/1",
  "kind": 3,
  "insertTextFormat": 2,
  "insertText": "parse_query(${1:A})",
  "data": {
    "module": "wtutil",
    "function": "parse_query",
    "arity": 1
  },
  "_emacsStartPoint": 892,
  "documentation": {
    "value": "## wtutil:parse_query/1\n\n---\n\n```erlang\n\n  parse_query(A) \n\n```\n\n```erlang\n-spec parse_query(#arg{}) -> parsed_query().\n```",
    "kind": "markdown"
  }
}
```

**Comment by yyoncho** [comment]
do you know if we have to resolve in order to get the tags?

**Comment by robertoaloi** [comment]
The specification doesn't seem to mention that. So, I would assume that _tags_ could be present already in the unresolved _CompletionItem_. In my case they are, and they are also part of the resolved item (the only difference between the unresolved and the resolved version of the item is the _documentation_ field).

**Comment by yyoncho** [comment]
That's good, thank you. Otherwise, we will have trouble because we would have to resolve synchronously all of the visible items.

**Comment by robertoaloi** [comment]
Makes sense, given that the plan is to show this via a _strikethrough_ directly in the dropdown.

**Comment by yyoncho** [comment]
Created PR #3227

**Comment by robertoaloi** [comment]
Brilliant, thanks!
